float64 distance
float64 angle
uint8 color
uint8 shape